pkg/controller/cdapmaster/cdapmaster_controller.go

```diff
@@ -389,15 +415,79 @@ func (b *Base) Observables(rsrc interface{}, labels map[string]string) []reconci
 
 // Objects for Messaging service
 func (s *Messaging) Objects(rsrc interface{}, rsrclabels map[string]string, observed, dependent, aggregated []reconciler.Object) ([]reconciler.Object, error) {
-	r := rsrc.(*alpha1.CDAPMaster)
-	return getStatefulServiceResources(&r.Spec.Messaging.CDAPStatefulServiceSpec, rsrc, rsrclabels, alpha1.ServiceMessaging)
+	master := rsrc.(*alpha1.CDAPMaster)
+	var upgradeJob *batchv1.Job
+	for _, item := range reconciler.ObjectsByType(observed, k8s.Type) {
+		if job, ok := item.Obj.(*k8s.Object).Obj.(*batchv1.Job); ok {
+			if master.Status.UpgradeJobStartTimeNano != 0 && strings.Contains(job.Name, strconv.FormatInt(master.Status.UpgradeJobStartTimeNano, 10)) {
+				upgradeJob = job
+				break
+			}
+		}
+	}
+	var upgradeResources []reconciler.Object
+	if upgradeJob == nil && master.Status.UpgradeJobStartTimeNano == 0 &&
```

Comment:
This function is kind of very big with multiple upgrade state handling inside. Please refactor them into multiple functions to make the intentions clear (launchJob, check Job, on Job finish).